src/android/device.rs

```diff
@@ -23,6 +23,102 @@ fn gradlew(config: &Config, env: &Env) -> bossy::Command {
+mod BundleToolJar {
+    use std::path::PathBuf;
+
+    use crate::android::device::{Config, Invalidutf8Error};
+
+    pub struct BundletoolJarInfo {
+        version: &'static str,
+    }
+
+    impl BundletoolJarInfo {
+        const fn new(version: &'static str) -> Self {
+            Self { version }
+        }
+
+        fn jar(&self) -> String {
+            format!("bundletool-all-{}.jar", self.version)
+        }
+
+        pub fn jar_path(&self, config: &Config) -> PathBuf {
+            config.project_dir().join(self.jar())
+        }
+
+        pub fn download_url(&self) -> String {
+            format!(
+                "https://github.com/google/bundletool/releases/download/{}/{}",
+                self.version,
+                self.jar()
+            )
+        }
+
+        pub fn run_command(&self, config: &Config) -> Result<String, Invalidutf8Error> {
+            let jar_path = self.jar_path(config);
+
+            Ok(format!(
+                "java -jar {}",
+                jar_path
+                    .to_str()
+                    .ok_or_else(|| Invalidutf8Error(jar_path.clone()))?
+            ))
+        }
+    }
+
+    pub const BUNDLE_TOOL_JAR_INFO: BundletoolJarInfo = BundletoolJarInfo::new("1.8.0");
+}
+
+fn bundletool_command_string(config: &Config) -> Result<String, Invalidutf8Error> {
+    #[cfg(not(target_os = "macos"))]
+    {
+        BundleToolJar::BUNDLE_TOOL_JAR_INFO.run_command(config)
+    }
+    #[cfg(target_os = "macos")]
+    {
+        Ok("bundletool".to_string())
+    }
+}
+
+fn install_bundletool(config: &Config) -> Result<(), BundletoolInstallError> {
+    #[cfg(not(target_os = "macos"))]
+    {
+        if !std::path::Path::new(&BundleToolJar::BUNDLE_TOOL_JAR_INFO.jar_path(config)).exists() {
+            let jar_path = BundleToolJar::BUNDLE_TOOL_JAR_INFO.jar_path(config);
+            let response = ureq::get(&BundleToolJar::BUNDLE_TOOL_JAR_INFO.download_url())
+                .call()
+                .map_err(BundletoolInstallError::DownloadFailed)?;
+            let mut out = std::fs::File::create(&jar_path).map_err(|err| {
+                BundletoolInstallError::JarFileCreationFailed {
+                    path: jar_path.clone(),
+                    cause: err,
+                }
+            })?;
+            std::io::copy(&mut response.into_reader(), &mut out).map_err(|err| {
+                BundletoolInstallError::CopyToFileFailed {
+                    path: jar_path,
+                    cause: err,
+                }
+            })?;
+        }
+    };
+    #[cfg(target_os = "macos")]
+    {
+        crate::apple::deps::install("bundletool", Default::default())
+            .map_err(BundletoolInstallError)?;
+    };
+    Ok(())
+}
+
+#[derive(Debug)]
+pub struct Invalidutf8Error(std::path::PathBuf);
+
+impl Reportable for Invalidutf8Error {
+    fn report(&self) -> Report {
+        Report::error("Path contained invaluid utf-8", format!("{:#?}", self.0))
```

Comment:
This isn't necessarily relevant after addressing the other comment above, but it's very important that the first arg of the report makes the context unambiguous. It's a high-level description of the general action that failed that the user can grasp (i.e. "APK installation failed") that also uniquely identifies (to us) where in the codebase that error came from.

The second arg then expands upon the first one with specific details, like the underlying error, relevant data (like the path here), etc. This is also ideally unambiguous and informative within the context provided by the first arg, so it's often important that i.e. the meaning of the path is included: `APK path {:?} contained invalid utf-8`

Your other `Reportable` impls look great though!